flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,238 @@
+
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> restored = Lists.newArrayList();
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+    long checkpointId = 0;
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+
+      if (request.getCommittable().checkpointId() > maxCommittedCheckpointId) {
+        if (request.getNumberOfRetries() > 0) {
+          restored.add(request);
+          dataFilesNumRestored = dataFilesNumRestored + committable.dataFiles().length;
+          deleteFilesNumRestored = deleteFilesNumRestored + committable.deleteFiles().length;
+        } else {
+          store.add(request);
+          dataFilesNumStored = dataFilesNumStored + committable.dataFiles().length;
+          deleteFilesNumStored = deleteFilesNumStored + committable.deleteFiles().length;
+        }
+        checkpointId = Math.max(checkpointId, request.getCommittable().checkpointId());
+      }
+    }
+
+    if (restored.size() > 0) {
+      commitResult(dataFilesNumRestored, deleteFilesNumRestored, restored);
+    }
+
+    commitResult(dataFilesNumStored, deleteFilesNumStored, store);
+    this.maxCommittedCheckpointId = checkpointId;
+  }
+
+  @Override
+  public void close() throws Exception {
+    if (tableLoader != null) {
+      tableLoader.close();
+    }
+
+    if (workerPool != null) {
+      workerPool.shutdown();
+    }
+  }
+
+  private void commitResult(int dataFilesNum, int deleteFilesNum,
+                            Collection<CommitRequest<FilesCommittable>> committableCollection) {
+    int totalFiles = dataFilesNum + deleteFilesNum;
+
+    if (totalFiles != 0) {
+      if (replacePartitions) {
+        replacePartitions(committableCollection, dataFilesNum, deleteFilesNum);
+      } else {
+        commitDeltaTxn(committableCollection, dataFilesNum, deleteFilesNum);
+      }
+    }
+  }
+
+  private void replacePartitions(Collection<CommitRequest<FilesCommittable>> writeResults, int dataFilesNum,
+                                 int deleteFilesNum) {
+    // Partition overwrite does not support delete files.
+    Preconditions.checkState(deleteFilesNum == 0, "Cannot overwrite partitions with delete files.");
+
+    // Commit the overwrite transaction.
+    ReplacePartitions dynamicOverwrite = table.newReplacePartitions().scanManifestsWith(workerPool);
+
+    String flinkJobId = null;
+    long checkpointId = -1;
+    for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+      try {
+        WriteResult result = commitRequest.getCommittable().committable();
+        Preconditions.checkState(result.referencedDataFiles().length == 0, "Should have no referenced data files.");
+
+        flinkJobId = commitRequest.getCommittable().jobID();
+        checkpointId = commitRequest.getCommittable().checkpointId();
```

Comment:
Are we sure that the `checkpointId` is monotonously increasing?

Reply:
yes , monotonically increasing in one taks.

Reply:
And it is monotonically increasing in our `for` loop as well?

Reply:
I think there are four types of checkpoint data in the for loop:
1. The data of the current checkpoint of the current job, not restored from the checkpoint-data.
2. The data of the current checkpoint of the current job, restored from the checkpoint-data.


Check whether the current checkpoint-id is larger than the checkpoint-id committed last time(From the memory cache, because the checkpointid of the current task is incremented.). If yes, continue to commite. If no, skip.

3. The data of the current checkpoint of the current job, but try to commit again

We determine the number of getNumberOfRetries. If the number is greater than 0, try to retry commit.

4. Pre job, must be recovered from the checkpoint-data.

Check whether the data of the checkpoint has been committed based on the jobID (from iceberg table metadata).


Therefore, the checkpoingid in the for loop is not strictly incremental. It contains the current checkpointid and restore checkpointid, which is come from the previous checkpoint of the current task(triggered when the job is automatically retried due to an error), and which is recovered from the previous job.

Reply:
In addition, if a job is restore from the previous job checkpoint, its checkpoint id will be incremented from the pre job.